app/controllers/components_controller.rb

```diff
@@ -2,15 +2,19 @@
 
 class ComponentsController < ApplicationController
   def show
-    @component = Component.get(params[:id])
+    @component = Component.get(params[:id]) || not_found
   end
 
   def index
     @components = Component.all
   end
 
   def preview
-    @component = Component.get(params[:component_id])
+    @component = Component.get(params[:component_id]) || not_found
     render layout: 'preview'
   end
+
+  def not_found
+    render status: 404, text: "404 error"
```

Comment:
You don't need to concern yourself with rendering here, you just need to return the correct status and let the application stack take care of everything else.

Additionally you are using the `||` operator for control flow within an assignment declaration which makes the code hard to reason about. The code reads *"set the component instance variable to the value returned by Component.get or if that is nil the value returned by calling the not_found method."* However, instead of returning a value the `not_found` method instead invokes a render.

```rb
def show
  @component = Component.get(params[:id])

  render :not_found if @component.nil?
end
```